File: backend/app/providers/traffic/factory.py
"""Selects the traffic provider from configuration."""

from __future__ import annotations

import logging
from functools import lru_cache

from app.core.config import settings
from app.providers.traffic.base import TrafficProvider
from app.providers.traffic.http import HttpTrafficProvider
from app.providers.traffic.mock import MockTrafficProvider

logger = logging.getLogger("roadwatch.traffic")


@lru_cache
def get_traffic_provider() -> TrafficProvider:
    provider = settings.TRAFFIC_PROVIDER.strip().lower()
    if provider == "http":
        return HttpTrafficProvider(settings.TRAFFIC_API_URL, settings.TRAFFIC_API_KEY)
    if provider != "mock":
        logger.warning("Unknown TRAFFIC_PROVIDER %r; using the development provider", provider)
    return MockTrafficProvider()

File: supabase/functions/api/providers/traffic/factory.ts
// Ported from backend/app/providers/traffic/factory.py.

import { settings } from "../../_shared/config.ts";
import type { TrafficProvider } from "./base.ts";
import { MockTrafficProvider } from "./mock.ts";
import { HttpTrafficProvider } from "./http.ts";

let cached: TrafficProvider | null = null;

export function getTrafficProvider(): TrafficProvider {
  if (cached) return cached;
  const provider = settings.TRAFFIC_PROVIDER.trim().toLowerCase();
  cached = provider === "http"
    ? new HttpTrafficProvider(settings.TRAFFIC_API_URL, settings.TRAFFIC_API_KEY)
    : new MockTrafficProvider();
  return cached;
}
